### PR TITLE
fix(rds): fixed the IAM policy that grantConnect() generates for DatabaseInstanceReadReplica

### LINKED ISSUE #31061: (aws-rds): grantConnect generates incorrect policy for DatabaseInstanceReadReplica
### Describe the bug

Calling `grantConnect` on an instance of `DatabaseInstanceReadReplica` generates an incorrect policy that uses the full ARN of the instance instead of the `instanceResourceId` value.

### Expected Behavior

```json
{
    "Action": "rds-db:connect",
    "Resource": "arn:aws:rds-db:us-east-1:1234567890:dbuser:db-INSTANCE_RESOURCE_ID/user",
    "Effect": "Allow"
}
```

### Current Behavior

```json
{
    "Action": "rds-db:connect",
    "Resource": "arn:aws:rds-db:us-east-1:1234567890:dbuser:arn:aws:rds:us-east-1:1234567890:db:instance-name-wq2y5qzlfdy6/user",
    "Effect": "Allow"
}
```

### Reproduction Steps

1. Create a read replica
2. Call grantConnect on it

### Possible Solution

_No response_

### Additional Information/Context

https://github.com/aws/aws-cdk/blob/abc78bfa613453185e59d5d9a17e5c5cfb5437b3/packages/aws-cdk-lib/aws-rds/lib/instance.ts#L201-L206

### CDK CLI Version

2.150.0

### Framework Version

2.150.0

### Node.js Version

v20.14.0

### OS

macOS

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ashishdhingra**
### Issue # (if applicable)

Closes #31061.

### Reason for this change
Calling `grantConnect()` on an instance of `DatabaseInstanceReadReplica` generates an incorrect policy that uses the full ARN of the instance instead of the instanceResourceId value. It should have created policy with correct resource format `arn:aws:rds-db:region:account-id:dbuser:DbiResourceId/db-user-name` per [Creating and using an IAM policy for IAM database access](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.IAMPolicy.html).

### Description of changes
Fixed the IAM policy that `grantConnect()` generates for `DatabaseInstanceReadReplica`. The change correctly sets the value of `instanceResourceId` to replica instance `attrDbiResourceId`. The value of `instanceResourceId` is used to generate IAM policy.

### Description of how you validated changes
- Added new unit test.
- Updated existing integration test.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts
Comment (moelasmar) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: I am afraid that this can be a breaking change to customers. I know that we assigned a wrong value, but may be there are some customers that use this property to get the DB Instance Arn and use it some how. So, my recommendation is to deprecate this property, and to add a new property for the instance resource id, and use it in grant method, and also check where else we use the current property.

**Review by moelasmar — DISMISSED** [review]
left one comment

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts
Comment (ashishdhingra) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: @moelasmar Thanks for your review comment. Would creating a feature flag (example https://github.com/aws/aws-cdk/pull/29794) be the recommended approach instead of adding a new property? Please advise.

Comment (moelasmar) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: I think the feature flag here will not fix the problem, as the flag default value will be false (to avoid breaking customers), so the policy will still referring to the wrong replica instance id. So, I do not prefer feature flag, it add a lot of confusion on us when we write code, so we can use this property in similar cases assuming that we are using the correct property, but because the feature flag is disabled, this may cause more issues for customers.

I prefer to go with deprecation, and adding new property.

Comment (ashishdhingra) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: @moelasmar Thanks for your review comment. I have added a commit to:
- Deprecate `instanceResourceId` property and add new `instanceResourceIdV2` property in `DatabaseInstanceReadReplica`.
- Added a copy of `grantConnect()` in `DatabaseInstanceReadReplica` that makes use of `instanceResourceIdV2`.

Please review.

Thanks,
Ashish

Comment (moelasmar) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: Thanks @ashishdhingra for resolving my comment, and sorry for late response. 

I actually, did a deeper check, and I found that I missed that the property `instanceResourceId` is actually exist in the Interface `IDatabaseInstance` so, adding a new property will actually break the idea of that interface, and also deprecating this property is not a good idea.

So, I think it is better to make this change under feature flag, and to enable the feature flag by default, and the customers who want to keep the current behaviour, they can disable the flag.

Comment (moelasmar) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: I pushed some updates to this change to add the feature flag. Could you please take a look.

Comment (ashishdhingra) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:1369]: Looks good reviewing at high level. Needs review from the maintainer.

File: packages/@aws-cdk/cx-api/FEATURE_FLAGS.md
Comment (paulhcsun) [inline packages/@aws-cdk/cx-api/FEATURE_FLAGS.md:1395]: The spacing on some of these seem off, was it manually generated or auto generated by running the build?

File: packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md
Comment (paulhcsun) [inline packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md:1396]: Do we want this to be `true` by default? Would that not break customers?

**Review by paulhcsun — COMMENTED** [review]
Changes look good, just a couple comments about how the feature flag is defined and the default value being `true`.

File: packages/@aws-cdk/cx-api/FEATURE_FLAGS.md
Comment (moelasmar) [reply packages/@aws-cdk/cx-api/FEATURE_FLAGS.md:1395]: it seems while I merged the conflicts I messed up these files, let me rerun the build and see if it will fix it

File: packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md
Comment (moelasmar) [reply packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md:1396]: my idea was which behaviour should I keep, shall I keep the wrong behaviour, or should I fix it. I chose to fix the wrong behaviour with an option to let people keep the current one. But, honestly you have a point, it is a breaking change

Comment (moelasmar) [reply packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md:1396]: I will change it to false.

**Comment by moelasmar** [comment]
for some reason I am not able to update this PR, and push changes to @ashishdhingra forked repo, so I created this new PR https://github.com/aws/aws-cdk/pull/31579 that includes @ashishdhingra changes, and close this one.

**Pull request closed without merging**
